src/find_sort_location.rs

```diff
@@ -0,0 +1,142 @@
+#![allow(dead_code)]
+
+use futures::future::join_all;
+
+use crate::{
+    field::Field,
+    helpers::ring::Ring,
+    replicated_secret_sharing::ReplicatedSecretSharing,
+    securemul::{ProtocolContext, SecureMul},
+};
+
+pub struct SecureSort<'a, R, F> {
+    context: &'a ProtocolContext<'a, R>,
+    share_of_one: ReplicatedSecretSharing<F>,
+}
+
+impl<F: Field, R: Ring> SecureSort<'_, R, F> {
+    pub async fn execute(
+        self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<ReplicatedSecretSharing<F>> {
+        let gen_permutation_share = |mult_output: Vec<ReplicatedSecretSharing<F>>| {
+            let len = mult_output.len() / 2;
+
+            // Fold the result
+            let mut permutation_share = Vec::new();
+            (0..len).for_each(|j| {
+                permutation_share.push(mult_output[j] + mult_output[j + len]);
+            });
+            permutation_share
+        };
+        gen_permutation_share(self.secure_multiply(self.prepare_mult_inputs(shares)).await)
+    }
+
+    fn prepare_mult_inputs(
+        &self,
+        shares: &[ReplicatedSecretSharing<F>],
+    ) -> Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)> {
+        // Step1: Calculate x and 1 - x
+        let mut x_and_one_minus_x: Vec<ReplicatedSecretSharing<F>> =
+            shares.iter().map(|x| self.share_of_one - *x).collect();
+        x_and_one_minus_x.append(&mut shares.to_owned());
+
+        // Step2: Calculate cumulative sum
+        let length = x_and_one_minus_x.len();
+        let mut cumulative_sum = Vec::with_capacity(length);
+        cumulative_sum.push(x_and_one_minus_x[0]);
+        for i in 1..length {
+            cumulative_sum.push(cumulative_sum[i - 1] + x_and_one_minus_x[i]);
+        }
+
+        // Step3: Consolidate step 1 and step 2 output
+        x_and_one_minus_x
```

Comment:
nit - you can consume `x_and_minus_x` at this moment so using `.into_iter()` will save an extra copy here

However, given that `x_and_one_minus_x.len() %2 == 0` you can also avoid an extra vector allocation here by chaining iterator calls


```rust
 fn prepare_mult_inputs(
        &self,
        shares: &[ReplicatedSecretSharing<F>],
    ) -> Vec<(ReplicatedSecretSharing<F>, ReplicatedSecretSharing<F>)> {
        // Step1: Calculate x and 1 - x
        let mut x_and_one_minus_x: Vec<ReplicatedSecretSharing<F>> =
            shares.iter().map(|x| self.share_of_one - *x).collect();
        x_and_one_minus_x.extend_from_slice(shares);

        x_and_one_minus_x
            // Step2: Calculate cumulative sum
            .windows(2)
            // SAFETY: input size is at least 2 so it is safe to assume slice always has 2 elements
            .map(|slice| slice[0] + slice[1])
            // Step3: Consolidate step 1 and step 2 output
            .zip(x_and_one_minus_x.iter())
            .map(|(cumulative_sum, x_and_one_minus)| (*x_and_one_minus, cumulative_sum))
            .collect()
    }

```

Reply:
what do you think of using `scan` to do this? https://github.com/martinthomson/raw-ipa/pull/76#discussion_r944655032

Reply:
that code is strictly better. we should use it instead

Reply:
It's not perfect.  You don't need to extend_from_slice if you are iterating, you can use Iterator::chain() instead.  Also, don't collect x_and_one_minus_x, just feed it into zip().

Reply:
@martinthomson what about my alternative? https://github.com/martinthomson/raw-ipa/pull/76#discussion_r944655032

Reply:
Yeah, that is better.  Still not sure about the use of .collect() on the end of that chain though.  I guess that this is a decision imposed by the API to the function though.